### ISSUE TITLE
Website Agent - JSON scraping analysis / JSON Parse Agent - JSON analysis

### ISSUE THREAD

**Issue by virtadpt** [huginn/huginn#2773]
Is there a way in either agent to do something like `number_of_objects($.results.in.a.hash.or.list)` or `{{ $results.in.a.hash.or.list | number_of_objects }}`?  If not, how difficult would it be to implement?

This seems like the sort of thing that there would be a Ruby gem for.

**Comment by dsander** [comment]
There is the [size filter](https://shopify.github.io/liquid/filters/size/), but I don't know if it works on hashes.

**Comment by virtadpt** [comment]
I tried that.  It does not.

**Comment by dsander** [comment]
Hmm how are you using the filter? It seems to work for me. This is the configuration of a `EventFormattingAgent`:

```json
{
  "instructions": {
    "size": "{{ hash | size }}"
  },
  "matchers": [],
  "mode": "clean"
}
```

I am sending this Event:

```json
{
  "hash": {
    "one": "??",
    "two": "??"
  }
}
```

And get this as a result:


```
[
  {
    "size": "2"
  }
]
```

**Comment by virtadpt** [comment]
I'm using it like this in an Event Formatting Agent to try to count the number of objects in a list:
```
{
  "instructions": {
    "hits": "{{ body.searchresults | size}}"
  },
  "mode": "clean"
}
```
Here's the data I'm working with, it's a set of search results from a Searx instance: https://pastebin.com/XGqR4i2u

Ultimately, what I'm trying to do is extract `{{ body.searchresults }}`, sort them based upon `{{ body.searchresults.*.score }}`, and then take either the first ten elements of the list or the entire list, depending upon which size is larger.

I can do `{{ body | size}}` and get `"hits": "28126"`, which is the number of bytes in the event (matching the value of the Content-Length HTTP header), but `{{ body.searchresults | size }}` always returns a 0.  I'm not sure if I need to clean my data up or not.

**Comment by virtadpt** [comment]
Okay, I've figured it out.  The Event Formatting Agent wasn't the right tool to use.  jq Agent was.  Here's what I did:
```
{
  "filter": "[.[\"body\"] | fromjson | .[\"results\"][] | {title: .title, url: .pretty_url, score: .score}] | sort_by(.score) | reverse | .[0:10]"
}
```
jq Agent ticket forthcoming.

**Comment by dsander** [comment]
Glad you figured it out!

> I can do `{{ body | size}}` and get `"hits": "28126"`, which is the number of bytes in the event (matching the value of the Content-Length HTTP header), but `{{ body.searchresults | size }}` always returns a 0. I'm not sure if I need to clean my data up or not.

That sounds like the contents of `body` is a string a not a object/hash. The `JsonParseAgent` could be an other option, but I don't know if Liquid would be able to handle the sorting.

**Comment by virtadpt** [comment]
That was the case.  "body":{} was escaped JSON - everything was \'d.  I had to use jq's `unescape` as a first step toward being able to access it.  That's a big part of what was messing with me.  Once I had that figured out, I figured that it'd be way easier to just use jq instead of a couple of other agents to process the data.

For future reference, here's how I used jq to process Searx search results.  I'll do a full writeup this weekend (with a PoC agent network).

```
{
  "filter": "[.[\"body\"] | fromjson | .[\"results\"][] | {title: .title, url: .pretty_url, score: .score}] | sort_by(.score) | reverse | .[0:10]"
}
```
